### PR TITLE
Method comparison: Reduce peak memory usage of image generation task

### PR DESCRIPTION
**Pull request by BenjaminBossan**
The training of the image generation benchmark is generally quite memory efficient, not exceeding 12 GB for most PEFT methods. However, there are memory peaks outside of training, during precomputation and eval, that can lead to peaks > 16 GB.

This was reported [here](https://github.com/huggingface/peft/issues/2310#issuecomment-4892723534) and I confirmed it locally: I ran the image benchmark with 8 GB of 24 GB of the GPU memory already allocated and indeed encountered an OOM error. **Edit**: The original poster confirmed that training completed with this branch.

There are two solutions in this PR to reduce memory usage:

1. Very simply don't move the transformer component to GPU before precomputing the latents. This is enough to not OOM during latent computation, but we still get OOM during eval.
2. Precompute the prompt embeddings. The prompts are all known in advance, hence we can precompute their embeddings. This allows us to completely remove the text encoder, which we would previously onload to GPU during evaluation. This change allows to avoid OOM during eval.

Note a subtlety: To ensure that Diffusers does not try to compute the negative prompt embeddings with the text encoder during evaluation, we compute the embedding for the empty string too.

With these changes applied, I could successfully train VeRA with 16 GB of VRAM. I compared the results with the [existing VeRA run](https://github.com/huggingface/peft/blob/74481e75366684f1afa2cb60d70cefa05ca2d2e3/method_comparison/image-gen/results/vera--flux2-klein-default.json) to ensure that they match:

train loss:

```
0.777 -> 0.777
0.761 -> 0.761
0.774 -> 0.774
0.766 -> 0.767
0.733 -> 0.733
0.763 -> 0.762
0.760 -> 0.759
```

memory reserved max:

`10565451776 -> 10645143552`

This is very close and I think it's confirmation enough that training works the same way as previously. The existing results are thus not invalidated and we don't need to rerun all experiments.

**Edit**

I included another change in this PR related to memory usage:

The transformer module is moved to the GPU late so that we can first precompute with, and then offload, other components without having a memory peak. This means that when `get_peft_model` is called, the transformer is still on CPU. However, some PEFT methods perform expensive operations during `get_peft_model`, e.g. calculating SVDs on the base weights. Those should not be run on CPU, as that would take ages.

The solution is to move the transformer to GPU before calling `get_peft_model` and then offload it again, before moving it again to GPU later. It's a bit wasteful but better than hitting a memory peak.

### CODE REVIEW

File: method_comparison/image-gen/run.py [hunk 1]
```diff
@@ -303,9 +341,20 @@ def train(
     eval_time = 0.0
     error_msg = ""
 
-    # pre-compute, since they don't change during training and we can keep the text encoder and VAE offloaded
-    prompt_embeds_cache, text_ids_cache = precompute_prompt_caches(
-        pipeline, train_dataset["prompts"], device_type, train_config=train_config
+    # pre-compute, since they don't change during training and we can keep the text encoder and VAE offloaded; the
+    # eval prompts are also encoded now, so that the text encoder is not needed anymore afterwards
+    eval_prompts = (
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply method_comparison/image-gen/run.py:346]: IMO no need to add this extra complexity just to save a tiny amount of runtime.

File: method_comparison/image-gen/run.py [hunk 2]
```diff
@@ -99,20 +99,45 @@ def update(self):
 
 
 def precompute_prompt_caches(
-    pipeline, prompts: list[str], device_type: str, train_config: TrainConfig
-) -> tuple[torch.Tensor, torch.Tensor]:
+    pipeline, train_prompts: list[str], eval_prompts: list[str], device_type: str, train_config: TrainConfig
+) -> tuple[torch.Tensor, torch.Tensor, dict[str, torch.Tensor]]:
+    """Precompue all prompt embeds needed during the run in a single pass over the text encoder.
+
+    This is mainly to save memory so that we don't need to either onload the text encoder during evaluation, which
```
Comment (githubnemo) [inline method_comparison/image-gen/run.py]:
```suggestion
    This is mainly to save memory so that we don't need to either unload the text encoder during evaluation, which
```

File: method_comparison/image-gen/run.py [hunk 3]
```diff
@@ -140,15 +165,19 @@ def precompute_latent_cache(
     return torch.cat(latents_cache, dim=0)
 
 
-def _generate_images(pipeline, *, generator, prompts: list[str], config: TrainConfig):
+def _generate_images(
+    pipeline, *, generator, prompts: list[str], prompt_cache: dict[str, torch.Tensor], config: TrainConfig
```
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply method_comparison/image-gen/run.py:169]: done

**Review by githubnemo — APPROVED** [review]
LGTM otherwise
